collector/logs/sources/tail/tailer.go

```diff
@@ -0,0 +1,201 @@
+package tail
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/Azure/adx-mon/collector/logs/engine"
+	"github.com/Azure/adx-mon/collector/logs/sources/tail/sourceparse"
+	"github.com/Azure/adx-mon/collector/logs/transforms/parser"
+	"github.com/Azure/adx-mon/collector/logs/types"
+	"github.com/Azure/adx-mon/pkg/logger"
+	"github.com/tenebris-tech/tail"
+	"golang.org/x/sync/errgroup"
+)
+
+type TailerConfig struct {
+	Target          FileTailTarget
+	UpdateChan      <-chan FileTailTarget
+	AckGenerator    func(*types.Log) func()
+	WorkerCreator   engine.WorkerCreatorFunc
+	CursorDirectory string
+	WorkerName      string
+}
+
+// Tailer is a specific instance of a file being tailed.
+type Tailer struct {
+	tail           *tail.Tail
+	shutdown       context.CancelFunc
+	errgroup       *errgroup.Group
+	database       string
+	table          string
+	logTypeParser  sourceparse.LogTypeParser
+	logLineParsers []parser.Parser
+	resources      map[string]interface{}
+}
+
+func StartTailing(config TailerConfig) (*Tailer, error) {
+	group, ctx := errgroup.WithContext(context.Background())
+	ctx, shutdown := context.WithCancel(ctx)
+
+	batchQueue := make(chan *types.Log, 512)
+	outputQueue := make(chan *types.LogBatch, 1)
+	tailConfig := tail.Config{Follow: true, ReOpen: true, Poll: true}
+	existingCursorPath := cursorPath(config.CursorDirectory, config.Target.FilePath)
+	fileId, position, err := readCursor(existingCursorPath)
+	if err == nil {
+		if logger.IsDebug() {
+			logger.Debugf("TailSource: found existing cursor for file %q: %s %d", config.Target.FilePath, fileId, position)
+		}
+		tailConfig.Location = &tail.SeekInfo{
+			Offset:         position,
+			Whence:         io.SeekStart,
+			FileIdentifier: fileId,
+		}
+	}
+
+	tailFile, err := tail.TailFile(config.Target.FilePath, tailConfig)
+	if err != nil {
```

Comment:
Just a comment and nothing to block the current PR. I wonder if we should consider taking inventory of the type of failure modes we can encounter at this point and try to distinguish between permanent and transitive, such that if the error is transitive we can attempt a retry. My concern is that, if my understanding of the code is correct, we essentially abandon tailing a file that returns an error for any reason, which means we in effect _lose_ those logs, especially if a rotation occurs.